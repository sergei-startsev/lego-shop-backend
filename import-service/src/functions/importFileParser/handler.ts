import { Handler, S3EventRecord } from 'aws-lambda';
import { S3Client, GetObjectCommand, CopyObjectCommand, DeleteObjectCommand } from '@aws-sdk/client-s3';
import csv from 'csv-parser';
import { Readable } from 'stream';
import pino from 'pino';

const importFileParser: Handler = async event => {
  const logger = pino();
  logger.info('importFileParser');
  logger.info(event);

  const record = event?.Records?.find((r: S3EventRecord) => r.eventName === 'ObjectCreated:Put');
  if (!record) {
    return event;
  }

  const bucket = record.s3.bucket.name;
  const key = record.s3.object.key;

  logger.info({ bucket });
  logger.info({ key });

  const client = new S3Client({ region: process.env.REGION });
  const getCommand = new GetObjectCommand({
    Bucket: bucket,
    Key: key
  });

  const response = await client.send(getCommand);
  if (!response?.Body) {
    return event;
  }

  await new Promise<void>((resolve, reject) => {
    (response.Body as Readable)
      .pipe(csv({ strict: true }))
      .on('data', data => {
        logger.info(data);
      })
      .on('error', reject)
      .on('end', () => {
        logger.info('importFileParser finished');
        resolve();
      });
  });

  const newKey = key.replace(/^uploaded/, 'parsed');
  const copyCommand = new CopyObjectCommand({
    CopySource: `${bucket}/${key}`,
    Bucket: bucket,
    Key: newKey
  });
  await client.send(copyCommand);
  logger.info(`${key} -> ${newKey}`);

  const deleteCommand = new DeleteObjectCommand({
    Bucket: bucket,
    Key: key
  });
  await client.send(deleteCommand);
  logger.info(`${key} was deleted`);

  return event;
};

export const main = importFileParser;
