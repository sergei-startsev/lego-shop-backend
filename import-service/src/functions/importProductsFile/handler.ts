import { Handler } from 'aws-lambda';
import { PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import createError from 'http-errors';
import * as z from 'zod';
import pino from 'pino';
import { serializeResponse } from '@libs/api-gateway';
import { middyfy } from '@libs/lambda';

const importProductsFile: Handler = async event => {
  const logger = pino();
  logger.info('importProductsFile');
  logger.info(event);

  const schema = z.object({
    name: z.string().min(3)
  });
  const parsed = schema.safeParse(event?.queryStringParameters);
  if (!parsed.success) {
    throw createError(400, parsed.error);
  }
  const { name } = parsed.data;

  const client = new S3Client({ region: process.env.REGION });
  const command = new PutObjectCommand({
    Bucket: process.env.BUCKET_NAME,
    Key: `uploaded/${name}`
  });
  const url = await getSignedUrl(client, command, { expiresIn: 3600 });

  return serializeResponse({ url });
};

export const main = middyfy(importProductsFile);
