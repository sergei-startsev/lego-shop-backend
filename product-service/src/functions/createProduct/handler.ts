import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, TransactWriteCommand } from '@aws-sdk/lib-dynamodb';
import { Handler } from 'aws-lambda';
import createError from 'http-errors';
import { v4 as uuidv4 } from 'uuid';
import * as z from 'zod';
import pino from 'pino';
import { serializeResponse } from '@libs/api-gateway';
import { middyfy } from '@libs/lambda';
import { Product, Stock } from '@customTypes/api-types';

const createProduct: Handler = async event => {
  const logger = pino();
  logger.info('createProduct');
  logger.info(event);

  const schema = z.object({
    title: z.string(),
    description: z.string(),
    price: z.number().positive(),
    count: z.number().nonnegative()
  });

  const parsed = schema.safeParse(event.body);
  if (!parsed.success) {
    throw createError(400, parsed.error);
  }

  const { title, description, price, count } = parsed.data;
  const id = uuidv4();
  const product: Omit<Product, 'count'> = { id, title, description, price };
  const stock: Stock = { product_id: id, count };

  const client = new DynamoDBClient({ region: process.env.REGION });
  const ddbDocClient = DynamoDBDocumentClient.from(client);
  const createProduct = new TransactWriteCommand({
    TransactItems: [
      {
        Put: {
          TableName: process.env.PRODUCTS_TABLE_NAME,
          Item: product
        }
      },
      {
        Put: {
          TableName: process.env.STOCK_TABLE_NAME,
          Item: stock
        }
      }
    ]
  });
  await ddbDocClient.send(createProduct);

  const output: Product = { id, title, description, price, count };
  return serializeResponse(output);
};

export const main = middyfy(createProduct);
