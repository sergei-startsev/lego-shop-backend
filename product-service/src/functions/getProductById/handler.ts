import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, QueryCommand } from '@aws-sdk/lib-dynamodb';
import { Handler } from 'aws-lambda';
import createError from 'http-errors';
import * as z from 'zod';
import pino from 'pino';
import { serializeResponse } from '@libs/api-gateway';
import { middyfy } from '@libs/lambda';
import { Product, Stock } from '@customTypes/api-types';

const getProductById: Handler = async event => {
  const logger = pino();
  logger.info('getProductById');
  logger.info(event);

  const schema = z.object({
    id: z.string().uuid()
  });
  const parsed = schema.safeParse(event?.pathParameters);
  if (!parsed.success) {
    throw createError(400, parsed.error);
  }
  const { id } = parsed.data;

  const client = new DynamoDBClient({ region: process.env.REGION });
  const ddbDocClient = DynamoDBDocumentClient.from(client);

  const getProductById = new QueryCommand({
    TableName: process.env.PRODUCTS_TABLE_NAME,
    KeyConditionExpression: 'id=:id',
    ExpressionAttributeValues: { ':id': id }
  });
  const products = (await ddbDocClient.send(getProductById))?.Items as Omit<Product, 'count'>[];
  const product = products?.[0];
  if (!product) {
    throw createError(404, 'Product not found');
  }

  const getCountById = new QueryCommand({
    TableName: process.env.STOCK_TABLE_NAME,
    KeyConditionExpression: 'product_id=:id',
    ExpressionAttributeValues: { ':id': id }
  });
  const stocks = (await ddbDocClient.send(getCountById))?.Items as Stock[];
  const count = stocks?.[0].count;

  return serializeResponse({ ...product, count });
};

export const main = middyfy(getProductById);
