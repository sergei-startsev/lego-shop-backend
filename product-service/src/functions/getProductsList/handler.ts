import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/api-gateway';
import { formatJSONResponse } from '@libs/api-gateway';
import mockData from '@mocks/data.json';

const getProductsList: ValidatedEventAPIGatewayProxyEvent<void> = async () => formatJSONResponse(mockData);

export const main = getProductsList;
