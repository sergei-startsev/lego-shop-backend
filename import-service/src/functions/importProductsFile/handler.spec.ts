import { Context } from 'aws-lambda';
import { main } from './handler';

jest.mock('@aws-sdk/s3-request-presigner', () => ({
  getSignedUrl: () => Promise.resolve('https://test-url.com')
}));

describe('importProductsFile', () => {
  const event = {
    version: '2.0',
    routeKey: 'GET /import',
    rawPath: '/import',
    rawQueryString: 'name=bar.csv',
    headers: {
      accept: '*/*',
      'accept-encoding': 'gzip, deflate, br'
    },
    queryStringParameters: {
      name: 'bar.csv'
    },
    requestContext: {},
    isBase64Encoded: false
  };

  beforeEach(() => {
    process.env.BUCKET_NAME = 'test_bucket_name';
  });

  afterEach(() => {
    process.env.BUCKET_NAME = undefined;
  });

  it('should return the signed url', async () => {
    const res = await main(event, {} as Context, () => {});
    expect(res).toMatchSnapshot();
  });

  it('should return 400 status code if name param is not provided', async () => {
    const newEvent = { ...event, queryStringParameters: {} };
    const res = await main(newEvent, {} as Context, () => {});
    expect(res).toMatchSnapshot();
  });

  it('should return 400 status code if name param does not match the schema', async () => {
    const newEvent = { ...event, queryStringParameters: { name: '1' } };
    const res = await main(newEvent, {} as Context, () => {});
    expect(res).toMatchSnapshot();
  });
});
