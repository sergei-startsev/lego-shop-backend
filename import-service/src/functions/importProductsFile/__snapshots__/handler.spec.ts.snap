// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importProductsFile should return 400 status code if name param does not match the schema 1`] = `
{
  "body": "[
  {
    "code": "too_small",
    "minimum": 3,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 3 character(s)",
    "path": [
      "name"
    ]
  }
]",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`importProductsFile should return 400 status code if name param is not provided 1`] = `
{
  "body": "[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "name"
    ],
    "message": "Required"
  }
]",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`importProductsFile should return the signed url 1`] = `
{
  "body": "{"url":"https://test-url.com"}",
  "statusCode": 200,
}
`;
